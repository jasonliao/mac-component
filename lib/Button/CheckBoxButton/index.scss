.checkbox-button {
  display: flex;
  align-items: center;
  cursor: default;

  .checkbox-button__box {
    display: inline-block;
    background: #fff;
    border-width: 1px;
    border-style: solid;
    border-color: #c8c8c8 #d7d7d7 #d2d2d2;
    border-radius: 3px;
    box-shadow: inset 0px 0.4px 0.1px 0px rgba(0, 0, 0, 0.2);
    box-sizing: border-box;
    margin-right: 10px;
    position: relative;
  }

  .checkbox-button__label {
    display: inline-block;
  }
}

.checkbox-button:active {
  .checkbox-button__box {
    background: #f0f0f0;
    border-color: #c3c3c3 #cacaca #ccc;
  }
}

.checkbox-button__checked {
  .checkbox-button__box {
    background: #3588fd;
    border: none;

    &::after {
      content: '\2713';
      color: #fff;
      font-size: 0.9em;
      position: absolute;
      left: 0.2em;
      top: 0.1em;
      transform: rotate(-10deg);
    }
  }
}

.checkbox-button__checked.checkbox-button__small {
  .checkbox-button__box::after {
    font-size: 0.88em;
  }
}

.checkbox-button__checked:active {
  .checkbox-button__box {
    background: #066afc;
    border: none;
  }
}

.checkbox-button.checkbox-button__disabled,
.checkbox-button.checkbox-button_disabled:active {
  .checkbox-button__box {
    background: #f6f6f6;
    border-color: #d7d7d7 #dedede #dbdbdb;
  }
  .checkbox-button__label {
    color: #bebebe;
  }
}

.checkbox-button__regular {
  font-size: 13px;
  .checkbox-button__box {
    width: 14px;
    height: 14px;
  }
}

.checkbox-button__small {
  font-size: 11px;
  .checkbox-button__box {
    width: 12px;
    height: 12px;
  }
}

.checkbox-button__mini {
  font-size: 9px;
  .checkbox-button__box {
    border-radius: 3px;
    width: 11px;
    height: 11px;
  }
}