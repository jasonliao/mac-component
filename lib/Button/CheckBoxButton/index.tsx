import * as React from 'react'
import classNames from 'classnames'

import './index.scss'

export interface CheckBoxButtonProps {
  size?: 'regular' | 'small' | 'mini'
  checked?: boolean
  position?: 'left' | 'right' | 'above' | 'below'
  style?: object
  className?: string
  disabled?: boolean
  onChange?: (checked: boolean, e: React.MouseEvent) => void
}

interface State {
  checked: boolean
}

export default class CheckBoxButton extends React.Component<CheckBoxButtonProps, State> {
  static defaultProps = {
    size: 'regular',
    position: 'left',
    style: {},
    className: '',
    disabled: false,
    onChange: () => {}
  }

  state = {
    checked: false
  }

  handleClick = (e: React.MouseEvent) => {
    const { disabled, onChange, checked } = this.props
    const hasChecked = this.hasChecked()

    if (!disabled) {
      if (checked === undefined) {
        this.setState({ checked: !hasChecked })
      }
      onChange(!hasChecked, e)
    }
  }

  hasChecked() {
    // 如果传了 `props.checked`，则使用
    // 没有则使用组件内 `state.checked`
    const { checked: stateChecked } = this.state
    const { checked } = this.props

    return checked === undefined ? stateChecked : checked 
  }

  render() {
    const { size, disabled, className, style, children } = this.props
    const hasChecked = this.hasChecked()

    return (
      <div className={
          classNames(
            className,
            'checkbox-button',
            `checkbox-button__${size}`,
            {
              'checkbox-button__checked': hasChecked,
              'checkbox-button__disabled': disabled
            }
          )
        }
        onClick={this.handleClick}
        style={style}
      >
        <i className="checkbox-button__box"></i>
        <span className="checkbox-button__label">{ children }</span>
      </div>
    )
  }
}