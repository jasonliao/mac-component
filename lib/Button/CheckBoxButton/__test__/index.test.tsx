import * as React from 'react'
import { shallow } from 'enzyme'
import * as renderer from 'react-test-renderer'

import CheckBoxButton, { CheckBoxButtonProps } from '../index'

describe('CheckBoxButton Component', () => {
  const label = 'Check'
  it('[UI] render label', () => {
    const wrapper = shallow(<CheckBoxButton>{ label }</CheckBoxButton>)

    const text = wrapper.find('.checkbox-button__label').text()
    expect(text).toBe(label)
  })

  it('[UI] render with mini size', () => {
    const props: CheckBoxButtonProps = { size: 'mini' }
    const wrapper = shallow(<CheckBoxButton { ...props } />)

    const hasClass = wrapper.find('.checkbox-button').hasClass(`checkbox-button__${props.size}`)
    expect(hasClass).toBe(true)
  })

  it('[UI] render disabled checkbox', () => {
    const props: CheckBoxButtonProps = { disabled: true }
    const wrapper = shallow(<CheckBoxButton { ...props } />)

    const hasClass = wrapper.find('.checkbox-button').hasClass('checkbox-button__disabled')
    expect(hasClass).toBe(true)
  })

  it('[BEHAVIOR] checkbox clicked(uncontrolled)', () => {
    const props: CheckBoxButtonProps = { onChange: jest.fn() }
    const wrapper = shallow(<CheckBoxButton { ...props } />)
    const stateChecked = wrapper.state('checked')

    wrapper.simulate('click')
    expect(props.onChange).toHaveBeenCalledTimes(1)
    expect(wrapper.state('checked')).toBe(!stateChecked)
  })

  it('[BEHAVIOR] disabled checkbox clicked(uncontrolled)', () => {
    const props: CheckBoxButtonProps = {
      disabled: true,
      onChange: jest.fn()
    }
    const wrapper = shallow(<CheckBoxButton { ...props } />)
    const stateChecked = wrapper.state('checked')

    wrapper.simulate('click')
    expect(props.onChange).toHaveBeenCalledTimes(0)
    expect(wrapper.state('checked')).toBe(stateChecked)
  })

  it('[BEHAVIOR] checkbox clicked(controlled)', () => {
    const props: CheckBoxButtonProps = {
      onChange: jest.fn(),
      checked: false
    }
    const wrapper = shallow(<CheckBoxButton { ...props } />)
    const stateChecked = wrapper.state('checked')

    wrapper.simulate('click')
    expect(props.onChange).toHaveBeenCalledTimes(1)
    expect(wrapper.state('checked')).toBe(stateChecked)
  })

  it('[BEHAVIOR] disabled checkbox clicked(controlled)', () => {
    const props: CheckBoxButtonProps = {
      onChange: jest.fn(),
      checked: false,
      disabled: true
    }
    const wrapper = shallow(<CheckBoxButton { ...props } />)
    const stateChecked = wrapper.state('checked')

    wrapper.simulate('click')
    expect(props.onChange).toHaveBeenCalledTimes(0)
    expect(wrapper.state('checked')).toBe(stateChecked)
  })

  it('[SNAPSHOT] match the snapshot', () => {
    const props: CheckBoxButtonProps = {
      size: 'small',
      checked: true,
      style: { width: '130px' },
      className: 'custom_classname'
    }
    const tree = renderer.create(<CheckBoxButton { ...props }>{ label }</CheckBoxButton>)
    expect(tree).toMatchSnapshot()

  })
})