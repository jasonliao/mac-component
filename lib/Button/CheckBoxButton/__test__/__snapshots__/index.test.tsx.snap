// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBoxButton Component [SNAPSHOT] match the snapshot 1`] = `
<div
  className="custom_classname checkbox-button checkbox-button__small checkbox-button__checked"
  onClick={[Function]}
  style={
    Object {
      "width": "130px",
    }
  }
>
  <i
    className="checkbox-button__box"
  />
  <span
    className="checkbox-button__label"
  >
    Check
  </span>
</div>
`;
